### PR TITLE
Implement new style of tanh

### PR DESCRIPTION
**Pull request by unnonouno**
I implemented new style of tanh. Users can implement grad of grad of LSTM with this function.

### CODE REVIEW

File: chainer/functions/activation/tanh.py [hunk 1]
```diff
@@ -24,37 +24,68 @@ def forward_cpu(self, x):
         y = utils.force_array(numpy.tanh(x[0]))
         self.retain_inputs(())
```
Comment (beam2d) [inline chainer/functions/activation/tanh.py]: This `retain_inputs` is not needed anymore.

File: chainer/functions/activation/tanh.py [hunk 2]
```diff
@@ -24,37 +24,68 @@ def forward_cpu(self, x):
...
 
     def forward_gpu(self, x):
         if chainer.should_use_cudnn('==always') and x[0].flags.c_contiguous:
             y = cudnn.activation_forward(x[0], _mode)
+            self.retain_inputs((0,))
+            self._use_cudnn = True
         else:
             y = cuda.cupy.empty_like(x[0])
             cuda.cupy.tanh(x[0], out=y)
             self.retain_inputs(())
```
Comment (beam2d) [inline chainer/functions/activation/tanh.py]: ditto

File: chainer/functions/activation/tanh.py [hunk 3]
```diff
@@ -24,37 +24,68 @@ def forward_cpu(self, x):
...
-    def backward_gpu(self, x, gy):
-        y = self.output_data[0]
+    def forward_gpu(self, inputs):
+        self.retain_inputs((0, 1))
+        y, gy = inputs
         if (chainer.should_use_cudnn('==always') and
-                x[0] is not None and x[0].flags.c_contiguous and
-                gy[0].flags.c_contiguous):
-            gx = cudnn.activation_backward(x[0], y, gy[0], _mode)
+                self.x is not None and self.x.flags.c_contiguous and
```
Comment (beam2d) [inline chainer/functions/activation/tanh.py]: The check on c-contiguity of `x` looks redundant.

File: chainer/functions/activation/tanh.py [hunk 4]
```diff
@@ -24,37 +24,68 @@ def forward_cpu(self, x):
...
                 'gx = gy * (1 - y * y)',
-                'tanh_bwd')(y, gy[0])
+                'tanh_bwd')(y, gy)
         return gx,
 
+    def backward(self, indexes, grad_outputs):
+        y, gy = self.get_retained_inputs()
+        g = grad_outputs[0]
+        grad_y = -2 * g * gy * y
+        ggy = g * (1 - y * y)
```
Comment (beam2d) [inline chainer/functions/activation/tanh.py]: How about this? The number of Function applications is reduced from 6 to 5.
```py
y_mul_g = y * g
grad_y = -2 * gy * y_mul_g
ggy = g - y * y_mul_g
```

File: tests/chainer_tests/training_tests/extensions_tests/test_computational_graph.py [hunk 5]
```diff
@@ -7,13 +7,25 @@
...
 from chainer import configuration
-from chainer import functions
+from chainer import function_node
 from chainer import links
 from chainer import testing
 from chainer import training
 from chainer.training.extensions import computational_graph as c
 
 
+class Function1(function_node.FunctionNode):
```
Comment (beam2d) [inline tests/chainer_tests/training_tests/extensions_tests/test_computational_graph.py]: Use `chainer.FunctionNode` in tests.

File: tests/chainer_tests/training_tests/extensions_tests/test_computational_graph.py [hunk 6]
```diff
@@ -7,13 +7,25 @@
...
+
+    def forward(self, inputs):
+        return inputs[0],
+
+
+class Function2(function_node.FunctionNode):
```
Comment (beam2d) [inline tests/chainer_tests/training_tests/extensions_tests/test_computational_graph.py]: ditto

File: chainer/functions/activation/tanh.py (hunk 3 above)
Comment (unnonouno) [reply chainer/functions/activation/tanh.py]: Why is it redundant?

**Comment by unnonouno** [comment]
i fixed them.

**Comment by beam2d** [comment]
Jenkins, test this please.

**Comment by beam2d** [comment]
Jenkins, test this please

**Comment by beam2d** [comment]
LGTM

**Pull request merged**